implement common consensus algorithms

noot opened the pull request:
**Summary of changes**
Changes introduced in this pull request:
- implement common consensus algorithms for `ProtocolStateChain` which is of type `Vec<ProtocolState>`

would love feedback on this in terms of  how to make my code more idiomatic and if the types I created for consensus make sense, in particular the traits and usage of generics, and if there's a better way to create a `Chain` type that's an ascending list of `ProtocolState`s. 

looking forward we will likely need to have a blocktree of all the current forks in the network, so maybe it would be nice to have a blocktree implementation and create chains from that.

**Reference issue to close (if applicable)**
<!-- Include the issue reference this pull request is connected to -->
<!--(e.g. Closes #1)-->
Closes #23 



<!-- Thank you 🔥 -->

creativcoder commented on types/src/protocol_state.rs:
```diff
@@ -0,0 +1,107 @@
+pub type Amount = u64;
+
+/// GlobalSlot contains (slot_number, slots_per_epoch)
+pub type GlobalSlot = (u64, u64);
```
Quite a few of these types are already defined in the base crate: https://github.com/ChainSafe/mina-rs/blob/main/base/src/global_slot.rs We can just import and use them.

noot replied:
wow I totally missed this, will update

creativcoder commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,88 @@
+        let mut hasher = Blake2b::new();
+        hasher.update(output);
+
+        // TODO: is there a prettier way to do this?
```
It would be good to add an impl for BaseHash from a slice:
```rust
// In crypto/hash/types.rs

impl<'a> From<&'a [u8]> for BaseHash {
    fn from(b: &'a [u8]) -> Self {
        let mut o = BaseHash::default();
        o.0.copy_from_slice(&b);
        o
    }
}

```
and then we can do:
```
let hash = hasher.finalize();
let hash = hash.as_slice(); // This temporary is just to make rust borrow checker happy
Some(BaseHash::from(hash))
````

creativcoder commented on base/src/protocol_state.rs:
```diff
@@ -39,20 +39,30 @@ pub struct ProtocolConstants {
+    fn get_height(&self) -> Length;
+}
+
+impl Header for ProtocolState {
```
Is `ProtocolState` the only implememter for `Header`?
What are other entities that can implement `Header`?

willemolding replied:
Even if no other types do implement it I like this for separation of concerns

noot replied:
@creativcoder right now it's just `ProtocolState` but potentially `ExternalTransition` could implement it as well

creativcoder commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,211 @@
+where
+    T: Header,
+{
+    fn push(&mut self, new: T) -> Result<(), &'static str>;
```
The second type param of Result needs to be a proper error type. We can define and enum ConsensusError which derives thiserror::Error from https://docs.rs/thiserror/1.0.26/thiserror/ crate use it in place of &'static str

willemolding replied:
Seconded, check out https://github.com/ChainSafe/serde-bin-prot/blob/main/src/error.rs for an example of writing a custom error type

noot replied:
updated!

creativcoder commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,211 @@
+            0 => (),
+            n => {
+                if new.get_height().0 != self.0[n - 1].get_height().0 + 1 {
+                    return Err("header must have height 1 greater than top");
```
can probably use an enum variant ConsensusError::InvalidHeight along with changes from the above comment.

noot replied:
updated

creativcoder commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    fn consensus_state(&self) -> Option<&ConsensusState> {
+        match self.top() {
```
We can replace the whole match with just `self.top().map(|s| &s.body.consensus_state)`

noot replied:
updated

creativcoder commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    fn global_slot(&self) -> Option<&GlobalSlot> {
+        match self.top() {
```
ditto as above

noot replied:
updated

willemolding commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,211 @@
+        BaseHash::from(hash.as_bytes()).to_hex()
+    }
+
+    fn state_hash(&self) -> Option<BaseHash> {
```
It is important to use the correct hash types. By convention Mina adds a prefix string to data before hashing it which is different depending on the type being hashes. Presumably this is to prevent a hash collision between different types with identical data. 

Check out the prefixes here https://github.com/ChainSafe/mina-rs/blob/8daf8e2a45e19c3b4912c688744faaa20cee6388/crypto/src/hash/prefixes.rs#L76

To avoid confusion I made the `Hash` and `Hashable` traits. Here `ProtocolState` should implement `Hashable` which returns a `StateHash` which implements `Hash`. This also makes these parts of the code much tidier since all the hashing logic will live in once place.

This is a bit tricky to describe in a comment so I'll link a branch showing how this could be done using our hashing lib

willemolding commented on crypto/src/hash/types.rs:
```diff
@@ -13,14 +13,15 @@
 
 #[derive(Default, Serialize, Deserialize, PartialEq, Debug, Clone)]
-pub(crate) struct BaseHash([u8; 32]);
+pub struct BaseHash([u8; 32]);
```
I'd be happier to keep this as in internal crate type if possible

willemolding commented on crypto/src/hash/types.rs:
```diff
@@ -31,10 +32,28 @@ impl From<HashBytes> for BaseHash {
     }
 }
 
+impl<'a> From<&'a [u8]> for BaseHash {
+    fn from(b: &'a [u8]) -> Self {
+        let mut o = BaseHash::default();
+        o.0.copy_from_slice(&b);
+        o
+    }
+}
+
+impl BaseHash {
+    pub fn to_hex(&self) -> String {
+        let mut s = String::from("0x");
+        for b in &self.0 {
+            write!(&mut s, "{:02x}", b).expect("failed to write hex byte")
+        }
+        s
+    }
+}
+
```
```suggestion
impl AsRef<[u8]> for BaseHash {
    fn as_ref(&self) -> &[u8] {
        &self.0
    }
}
```

and then if we use the `hex` crate this type will automatically implement `to_hex` whenever the trait is in scope!

noot replied:
oh nice, so by implementing `AsRef` for a type, then  `to_hex` automatically gets implemented?

willemolding replied:
Yep you got it As long as the trait is in scope that is

willemolding reviewed (changes requested):
Amazing job overall. Stoked to be making progress on the next milestone.

A few changes suggested around the use of the hashing lib. I've left some comments in here but to make it easier I have a branch with some of the changes suggested which I think will be much easier to follow.

It also required some changes to the hash lib so it was great to have that tested out. 

Some suggested changes are in here https://github.com/ChainSafe/mina-rs/tree/noot/common-willem-rev

creativcoder commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,211 @@
+            .map(|s| (s.slot_number.0 % s.slots_per_epoch.0))
+    }
+
+    fn length(&self) -> u64 {
```
Better to use a `usize` type as it's a generic type which changes based on architecture. On a 64 bit it'll be a u64 and likewise for 32bit.

noot replied:
done

creativcoder commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    fn state_hash(&self) -> Option<BaseHash> {
+        let s = match self.top() {
```
It would be great to use if let Some(s) = self.top() here as match is best applicable for more than 3 match arms.

noot replied:
done, I used an else case to return `None`:
```
        if let Some(s) = self.top() {
            let mut output = Vec::<u8>::new();
            to_writer(&mut output, &s).unwrap();
            let hash = blake2b(32, &[], &output);
            Some(BaseHash::from(hash.as_bytes()))
        } else {
            None
        }
```

is this the right way to do this?

willemolding replied:
You could use `map` here as well since you are only changing what is contained in the `Some` variant
```rust
self.top().map(|s| {
    let mut output = Vec::<u8>::new();
    to_writer(&mut output, &s).unwrap();
    let hash = blake2b(32, &[], &output);
    BaseHash::from(hash.as_bytes())
})
```

willemolding replied:
DW looks like you are all over this

creativcoder commented on consensus/src/common.rs:
```diff
@@ -0,0 +1,211 @@
+    fn last_vrf(&self) -> String {
+        let s = match self.top() {
+            Some(s) => s,
+            None => return "0x".to_string(),
```
Is this correct? I guess we should return a `None`, if the the chain is empty.

noot replied:
updated to return `Option<String>`

creativcoder reviewed (changes requested):
Thank you for giving this a head start! Left a few comments

noot commented:
@willemolding thanks for the updates on your branch, I pulled them into here, much cleaner to use the `Hashable` trait :) 

@creativcoder addressed your comments!

willemolding commented on crypto/src/hash/traits.rs:
```diff
@@ -13,7 +13,7 @@ const BLAKE_HASH_SIZE: usize = 32;
 pub trait Hash
 where
-    Self: From<Box<[u8]>>,
+    Self: From<Box<[u8]>> + AsRef<[u8]>,
```
Does this need an implementation of `AsRef` as a trait bound? It would be nice if we could drop this from here so that types can still implement `Hash` either way and just need to implement `AsRef<[u8]>` if they need converting to hex.

noot replied:
removed

willemolding reviewed (approved):
Looks amazing! One tiny comment but that's all

creativcoder commented on crypto/src/hash/traits.rs:
```diff
@@ -60,6 +60,12 @@ mod tests {
         const PREFIX: &'static HashPrefix = PROTOCOL_STATE;
     }
 
+    impl AsRef<[u8]> for TestHash {
```
I guess TestHash was for testing and can be removed?

noot replied:
it's used in the test cases here, so would be good to keep it

creativcoder commented on crypto/src/hash/types.rs:
```diff
@@ -31,10 +30,24 @@ impl From<HashBytes> for BaseHash {
     }
 }
 
+impl<'a> From<&'a [u8]> for BaseHash {
```
Do we still need this impl, after implementing `AsRef` below?

noot replied:
I think the other hash implementations still need it, for example this doesn't work if there's no `AsRef` for `BaseHash`:
```
pub struct StateHash(BaseHash);

impl AsRef<[u8]> for StateHash {
    fn as_ref(&self) -> &[u8] {
        &self.0.as_ref()
    }
}
```

creativcoder reviewed (approved):
Neat! let's merge this soon post last set of changes.
